lib/opal/nodes/literal.rb

```diff
@@ -213,42 +213,132 @@ class XStringNode < Base
       handle :xstr
 
       def compile
-        children.each do |child|
-          case child.type
-          when :str
-            value = child.loc.expression.source
-            if expr? && children.size == 1
-              value = detect_and_remove_trailing_semicolon(value, child)
+        should_add_semicolon = false
+        has_embeded_return   = false
+        returning, children  = unpack_return(self.children)
+        stripped_children    = strip_empty_children(children)
+        single_line          = single_line?(stripped_children)
+        single_child         = stripped_children.size == 1
+
+        if single_line
+          # If it's a single line we'll try to:
+          #
+          # - strip empty lines
+          # - remove a trailing `;`
+          # - detect an embedded `return`
+          # - prepend a `return` when needed
+          # - append a `;` when needed
+          # - warn the user not to use the semicolon in single-line x-strings
+
+          first_child = stripped_children.shift
+
+          if first_child.type == :str
+            first_value = first_child.loc.expression.source.strip
+            has_embeded_return = first_value =~ /^return\b/
+          end
+
+          if returning && !has_embeded_return
+            push('return ')
+          end
+
+          last_child = stripped_children.pop || first_child
+          if last_child.type == :str
+            last_value = last_child.loc.expression.source.rstrip
+            if (returning || expr?) && last_value[-1] == ';'
+              compiler.warning(
+                'Removed semicolon ending x-string expression, interpreted as unintentional',
+                last_child.line,
+              )
+              last_value = last_value[0...-1]
```

Comment:
Also maybe it's better to rewrite it to `[0..-2]`. Initially I thought that there's a bug (or a weird self-wriiten `dup` 😄 )